04-onion-routing.md

```diff
@@ -367,9 +378,60 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Trampoline Payments
+
+Trampoline payments allow nodes with an incomplete view of the network to
+delegate the construction of parts of the route to trampoline nodes.
+
+The origin node only needs to select a set of trampoline nodes and to know a
+route to the first trampoline node. Each trampoline node is responsible for
+finding its own route to the next trampoline node. The last trampoline node
+must be the final recipient.
+
+The `trampoline_onion_packet` has a variable size to allow implementations to
+choose their own trade-off between flexibility and privacy. It's recommended to
+add trailing filler data to the `trampoline_onion_packet` when using a small
+number of hops. It uses the same onion construction as the `onion_packet` and
+is embedded inside an `onion_packet`.
+
+Trampoline nodes are free to use as many hops as they want between themselves
+as long as they are able to create a route that satisfies the `cltv` and `fees`
+requirements contained in the onion.
+
+#### Requirements
+
+A sending node:
+
+- If the invoice doesn't support the `trampoline_routing` feature:
+  - MUST NOT use trampoline routing to pay that invoice
+- MUST ensure that each hop in the `trampoline_onion_packet` supports `trampoline_routing`
+- MUST encrypt the `trampoline_onion_packet` with the same construction as `onion_packet`
+- MAY add trailing filler data similar to what is done in the `onion_packet`
+- MUST use a different `session_key` for the `trampoline_onion_packet` and the `onion_packet`
+- MUST include the `trampoline_onion_packet` tlv in the _last_ hop's payload of the `onion_packet`
+- MUST include the invoice's `payment_secret` in the _last_ hop's payload of the `trampoline_onion_packet`
+- MUST generate a different `payment_secret` to use in the outer onion
```

Comment:
Why include payment_secret at all in the outer onion?

Reply:
It's for the case where trampoline nodes aggregate incoming MPP and then re-split differently to reach the next trampoline node.
This ensures they use the normal MPP validation code and no intermediate nodes can cheat. They could just rely on the `total_amount` to verify they receive everything, but I like the fact that it works just like normal payments.

Reply:
Ah, you need total_msat from payment_data.  But payment_secret here doesn't help much: the aggregating node can't know what the correct value is (and I can't see where in the spec you say it should be checked...).

An intermediate can probe by sending its own partial payment with some random payment_secret, so you can't really even say "they must all be the same...".

Reply:
It's true that in the invoice-based flow the recipient generated the `payment_secret` and can check the correctness of the value. In the trampoline flow, it's the trampoline sender who generates one, and the recipient blindly accepts and just uses it to bundle together HTLCs from the same set.

But I think it still has some value because it does ensure that HTLCs from two different nodes will not end up being bundled together (and mess up the payment). If we have the following payment (with trampoline route `Alice -> T1 -> T2 -> Bob`):

```
Alice -----> ... -----> T1 -----> I1 -----> I2 -----> T2 -----> ... -----> Bob
```

`T1` generated a `payment_secret` to send HTLCs to `T2`. `I2` is trying to interfere by generating his own (trampoline) payment to `T2`. They will have a different `payment_secret`, so they won't interfere with the ones `T1` sent.